Add support for battery-backed RAM for STM32 deivces

duda-patryk opened the pull request:
Hi!

These changes introduce BBRAM support for STM32 devices. For now, only backup registers from RTC are supported.
I also implemented bbram shell command which is convenient tool to examine and modify BBRAM contents.
Any remarks and comments are welcome (especially with regard to DTS).

Best regards,
Patryk

duda-patryk committed:
drivers: bbram: Introduce STM32 BBRAM driver

STM32 battery-backed RAM is organized in 4 byte registers. Number of
registers can vary between models from 5 to 32 registers.

Usually, the registers are part of RTC. On some variants they are part
of tamper module. On STM32F1 the registers are in separate module. For
now, only backup registers from RTC are supported.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
drivers: bbram: Introduce BBRAM shell

Add 'bbram' command which is very convenient tool for examining and
modyfying BBRAM content.

Examples:

Write one byte:
$ bbram write backup_regs 0x1 0xaa

Write many bytes (starting from 0x13):
$ bbram write backup_regs 0x13 0xa 0xb 0xc 0xd 0xe 0xf 0xaa 0xab 0xac

Read whole BBRAM:
$ bbram read backup_regs

Read many bytes (4 bytes starting from address 0x14 in this case):
$ bbram read backup_regs 0x14 4

Read one byte (from 0x14 address):
$ bbram read backup_regs 0x14

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: h7: Add RTC BBRAM to DTS

Add BBRAM entry for all STM32H7 microcontrollers except STM32H7A3 and
STM32H7B3 which have backup registers in tamper module.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: f4: Add RTC BBRAM to DTS

All STM32F4 devices have 20 4-byte battery-backed backup registers in
RTC.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: wb: Add RTC BBRAM to DTS

All STM32WB devices have 20 4-byte battery-backed backup registers in
RTC.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: l4: Add RTC BBRAM to DTS

STM32L4 devices, except STM32L412 STM32L422 STM32L4P5 STM32L4Q5, have
32 4-byte battery-backed RTC backup registers. Other STM32L4 devices
have backup registers in tamper module, not used in Zephyr.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: l1: Add RTC BBRAM to DTS

All STM32L1 devices have 4-byte battery-backed RTC registers, but
they have different number of registers:

STM32L151Xb-a has 5 registers.
STM32L151Xb has 20 registers.
STM32L151Xc, STM32L152Xc, STM32L152Xe have 32 registers.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: f3: Add RTC BBRAM to DTS

All supported STM32F3 devices have 4-byte battery-backed RTC registers,
but they have different number of registers:

STM32F303x8 and STM32F334 have 5 registers.
STM32F303xc and STM32F303xe have 16 registers.
STM32F302x8 has 20 registers.
STM32F373 has 32 registers.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: f7: Add RTC BBRAM to DTS

All STM32F7 devices have 32 4-byte battery-backed RTC registers.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: f0: Add RTC BBRAM to DTS

Among supported devices only STM32F031, STM32F051, STM32F072 and
STM32f09x have 5 4-byte battery-backed RTC registers.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: l0: Add RTC BBRAM to DTS

All STM32L0 devices have 5 4-byte battery-backed RTC registers.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
dts: arm: st: f2: Add RTC BBRAM to DTS

All STM32F2 devices have 20 4-byte battery-backed RTC registers.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

yperess commented on drivers/bbram/bbram_shell.c:
```diff
@@ -0,0 +1,206 @@
+/* Buffer is only needed for bytes that follow command, device and address. */
+#define BUF_ARRAY_CNT (CONFIG_SHELL_ARGC_MAX - 3)
+
+static int parse_ul(const char *str, unsigned long *result)
+{
+	char *end;
+	unsigned long val;
+
+	val = strtoul(str, &end, 0);
+
+	if (*str == '\0' || *end != '\0') {
+		return -EINVAL;
+	}
+
+	*result = val;
+	return 0;
+}
+
+static int parse_u32(const char *str, uint32_t *result)
+{
+	unsigned long val;
+
+	if (parse_ul(str, &val) || val > 0xffffffff) {
+		return -EINVAL;
+	}
+	*result = (uint32_t)val;
+	return 0;
+}
+
+static int parse_u8(const char *str, uint8_t *result)
+{
+	unsigned long val;
+
+	if (parse_ul(str, &val) || val > 0xff) {
+		return -EINVAL;
+	}
+	*result = (uint8_t)val;
+	return 0;
+}
```
Why do we need these? `stdlib` includes `strtoul`, can we use that?

semihalf-duda-patryk replied:
Actually, `strtoul` is used in line 25, these functions are a wrapper for `strtoul` call with error handling.
Function `parse_ul` checks `end` variable, so we know if there was a problem with parsing.
Functions `parse_u32` and `parse_u8` check if number is in expected range.

semihalf-duda-patryk replied:
Added.

semihalf-duda-patryk commented on drivers/bbram/bbram_shell.c:
```diff
@@ -0,0 +1,206 @@
+#include <zephyr/shell/shell.h>
+#include <zephyr/sys/util.h>
+
+#include <stdlib.h>
+#include <string.h>
```
Done.

GeorgeCGV commented on drivers/bbram/bbram_stm32.c:
```diff
@@ -0,0 +1,103 @@
+		return -EFAULT;
+	}
+
+	for (size_t written = 0; written < size; written += to_copy) {
```
`to_copy` value is indeterminate.

GeorgeCGV commented on drivers/bbram/bbram_stm32.c:
```diff
@@ -0,0 +1,103 @@
+		return -EFAULT;
+	}
+
+	for (size_t read = 0; read < size; read += to_copy) {
```
`to_copy` value is indeterminate.

semihalf-duda-patryk replied:
In line 40 we assign number of bytes we want to copy from register, so `read += to_copy` in line 37 is valid.

GeorgeCGV replied:
@semihalf-duda-patryk , what about the first iteration? It is not guaranteed that `to_copy` is `0`. Therefore, `read += to_copy` might end up with `read` being invalid.

semihalf-duda-patryk replied:
The first iteration is also correct. The third statement in for loop (`read += to_copy`) is executed at the end of iteration (after the loop body).

GeorgeCGV replied:
indeed, my bad; sorry

yperess reviewed (changes requested):
Please add `inline` to the static functions that do the parsing

semihalf-duda-patryk commented:
Added `inline` to the functions that do the parsing.

erwango commented:
I will review this week

gmarull reviewed (dismissed):
This is already supported https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/boards/stm32/backup_sram

semihalf-duda-patryk commented:
> This is already supported https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/boards/stm32/backup_sram

No. Backup SRAM is different than RTC backup RAM. Let's take a look at STM32H7 (RM0433 reference manual):
- The manual mentions about Backup SRAM in `2.4 Embedded SRAM` chapter:
```
The backup RAM is mapped at the address 0x3880 0000 and is accessible by most of the
system masters through D3 domain’s AHB matrix. With a battery connected to the V BAT pin,
the backup SRAM can be used to retain data during low-power mode (Standby and VBAT
mode).
```
- RTC backup RAM appears in `46.3.1 RTC block diagram`
```
The backup registers (RTC_BKPxR) are implemented in the RTC domain that
remains powered-on by VBAT when the VDD power is switched off
```

STM32H7 has 32 4-byte RTC backup RAM registers and 4Kbytes Backup SRAM.

erwango commented on drivers/bbram/Kconfig.stm32:
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2022 Google Inc
```
Just to verify and since provided mail differs: Is that accurate ?

semihalf-duda-patryk replied:
Yes

erwango commented on drivers/bbram/Kconfig:
```diff
@@ -14,6 +14,7 @@ source "subsys/logging/Kconfig.template.log_config"
 
 config BBRAM_INIT_PRIORITY
 	int "Init priority"
+	default 65 if BBRAM_STM32
```
Would you mind adding a comment on this choice (in commit message for instance)

semihalf-duda-patryk replied:
Added comment in a file.

erwango commented on dts/arm/st/f0/stm32f031.dtsi:
```diff
@@ -28,5 +28,10 @@
 				#pwm-cells = <3>;
 			};
 		};
+		bbram: bbram@40002850 {
+			compatible = "st,stm32-bbram";
+			reg = <0x40002850 0x14>;
+			status = "disabled";
+		};
```
Rather than adding dedicated nodes and adding a dependency on `COUNTER_RTC_STM32`, I'd prefer this is added as a child of existing RTC nodes, in the same way we're proceeding with PWM and timers and in lin with with is proposed in this [new RTC API proposal](https://github.com/zephyrproject-rtos/zephyr/pull/52618)

semihalf-duda-patryk replied:
Done. Please check if it is correct now.

erwango reviewed (changes requested):
Aside the device tree part, please add a way to get this compiled through `tests/drivers/bbram`.

semihalf-duda-patryk commented:
Addressing reviewer comments. Added way to compile this through `tests/drivers/bbram` and tested if it's really compiled.

erwango commented on tests/drivers/bbram/testcase.yaml:
```diff
@@ -13,3 +13,8 @@ tests:
     extra_args: CONF_FILE="npcx.conf"
     filter: dt_compat_enabled("nuvoton,npcx-bbram")
+  drivers.bbram.stm32:
+    platform_allow: stm32f0_disco stm32f3_disco stm32f4_disco stm32f723e_disco
+                    stm32h735g_disco stm32l1_disco stm32l476g_disco
+                    nucleo_wb55rg
```
When technically possible, avoid use of `platform_allow`.
In this case, why not enabling bbram on those boards and filter on `dt_compat_enabled()` as done on other platforrms ?

semihalf-duda-patryk replied:
> When technically possible, avoid use of platform_allow.

OK. Understood.

> In this case, why not enabling bbram on those boards and filter on dt_compat_enabled() as done on other platforrms ?

Done

erwango commented on tests/drivers/bbram/stm32.conf:
```diff
@@ -0,0 +1,7 @@
+# Copyright 2022 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_ZTEST=y
+CONFIG_ZTEST_NEW_API=y
+CONFIG_COUNTER=y
+CONFIG_BBRAM=y
```
Not blocking (not related to your PR): This test case implementation should eventually be refactored to use a common prj.conf.
Created https://github.com/zephyrproject-rtos/zephyr/issues/52754 to get it fix.

erwango commented on drivers/bbram/Kconfig.stm32:
```diff
@@ -0,0 +1,16 @@
+config BBRAM_STM32
+	bool "ST STM32 Battery-backed RAM drivers"
+	default y
+	depends on DT_HAS_ST_STM32_BBRAM_ENABLED
```
```
	depends on COUNTER
```

semihalf-duda-patryk replied:
Done

erwango commented on drivers/bbram/bbram_stm32.c:
```diff
@@ -0,0 +1,107 @@
+{
+	const struct bbram_stm32_config *config = dev->config;
+
+	if (!device_is_ready(config->parent)) {
+		LOG_ERR("Device %s is not ready", config->parent->name);
+		return -ENODEV;
+	}
```
I'd like to open the discussion on this part of your PR which seems the most controversial. This isn't a blocking comment.

At first sight, this dependency will save few bytes of ROM when both COUNTER and BBRAM are required by the application. On the other hand this will waste ROM in the case only BBRAM is required by the application, as COUNTER has to be complied even if not used.

For stm32 counter_ll_timer and pwm drivers, which share a similar constraint of using same IP for 2 different drivers, it was chose to make them fully independent. This is a slightly different case as in this case their usage is mutually exclusive on the same instance which is not true on current case.

Ideally we should be able to use multi function devices in that case. But I'm not sure there is any example yet available so I won't block.

semihalf-duda-patryk replied:
> For stm32 counter_ll_timer and pwm drivers, which share a similar constraint of using same IP for 2 different drivers, it was chose to make them fully independent. This is a slightly different case as in this case their usage is mutually exclusive on the same instance which is not true on current case.

We could add parent node for counter and backup_regs which would be responsible for basic initialization (clocks, interrupts), but it would be unnatural in my opinion.

It would be nice if we could implement both API (rtc and bbram) in one driver, with one init function, but I don't know if it's possible.

duda-patryk committed:
boards: arm: Enable RTC backup RAM on some boards

This patch enables STM32 RTC BBRAM in DTS on some STM32 boards for
testing purposes.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

duda-patryk committed:
tests: drivers: bbram: Test BBRAM on STM32 devices

This patch makes possible to check if STM32 boards compile with BBRAM
support enabled.

Signed-off-by: Patryk Duda <pdk@semihalf.com>

semihalf-duda-patryk commented:
Enabled bbram on some STM32 boards. Confirmed that the driver is built without problems. Added `depends on COUNTER`

erwango commented on dts/bindings/memory-controllers/st,stm32-bbram.yaml:
```diff
@@ -0,0 +1,20 @@
+properties:
+    st,backup-regs:
+        type: int
+        required: true
```
Please add a meaningful description.

erwango reviewed (changes requested):
Otherwise LGTM

semihalf-duda-patryk commented:
Rebased and fixed merge conflict
